VCU118 FPGA Updates + FireMarshal on Prototypes (#849)
ucb-bar/chipyard: docs/Prototyping/VCU118.rst

.. Note:: Remember that since whenever a new test harness is created (or the config changes, or the config packages changes, or...), you need to modify the make invocation.
    For example, ``make SUB_PROJECT=vcu118 CONFIG=MyNewVCU118Config CONFIG_PACKAGE=this.is.my.scala.package bitstream``.
    See :ref:`Prototyping/General:Generating a Bitstream` for information on the various make variables.

> Running Linux with Basic and Bringup Platforms

Author: The overall section splits the VCU118 into 2 sections the "Basic Design" and "Platform", I can modify this a bit to be "Basic Design" and "Bringup Design" and change this title to be clearer.

----------------------------------------------

As mentioned above, the default VCU118 harness is setup with a UART and a SPI SDCard.
These are utilized to both interact with the DUT (with the UART) and load in Linux (with the SDCard).
The following steps describe how to build and run buildroot Linux on the prototype platform.

Building Linux with FireMarshal
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> Since the prototype does not have a block device, we build Linux with the rootfs built into the binary (otherwise known as "initramfs" or "nodisk" version of Linux).

Author: I assume it probably can, but I don't know how to do that (nor do I think it will make much of a difference). Having it as a block device would change things in a couple ways:

1. Have to figure out how to get it to work with Linux / SPI interface. Create new partitions? Have a block device driver? Etc.
2. Could probably minimize the binary size. Thus making loading faster if we change the bootrom.

To make building this type of Linux binary easy, we will use the FireMarshal platform (see :ref:`fire-marshal` for more information).

1. Setup FireMarshal (see :ref:`fire-marshal` on the initial setup).
2. By default, FireMarshal is setup to work with FireSim.
   Instead, we want to target the prototype platform.
   This is done by switching the FireMarshal "board" from "firechip" to "prototype" using ``marshal-config.yaml``:

.. code-block:: shell

>     echo "board-dir : 'boards/prototype'" > $PATH_TO_FIREMARSHAL/marshal-config.yaml

Author: This file doesn't exist unless the user makes it. So this is assuming we create it from scratch.

.. Note:: Refer to the FireMarshal docs on more ways to set the board differently through environment variables and more.

3. Next, build the workload (a.k.a buildroot Linux) with nodisk with FireMarshal.
   For the rest of these steps, we will assume you are using the base ``br-base.json`` workload.

>    This workload has basic support for GPIO and SPI drivers but you can build off it in different workloads (refer to FireMarshal docs on workload inheritance).

Author: It does nothing special with the UART stuff... so its the default UART driver.

.. code-block:: shell

    ./marshal -v -d build br-base.json # here the -d indicates --nodisk or initramfs

.. Note:: Using the "board" FireMarshal functionality allows any child workload depending on ``br-base.json`` to use the "prototype" ``br-base.json`` rather than the FireChip version.
   Thus, you can re-use existing workloads that depend on ``br-base.json`` on the prototype platform by just changing the "board"!

4. The last step to generate the proper binary is to flatten it.
   This is done by using FireMarshal's install feature and will produce a ``*-flat`` binary in the ``$PATH_TO_FIREMARSHAL/images`` directory (in our case ``br-base-bin-nodisk-flat``).

.. code-block:: shell

   ./marshal -v -d install -t prototype br-base.json

Setting up the SDCard
~~~~~~~~~~~~~~~~~~~~~

These instructions assume that you have a spare uSDCard that can be loaded with Linux and other files using two partitions.

> The 1st partition will be used to store the Linux binary (created with FireMarshal or other means) while the 2nd partition will be used to store miscellaneous files.

Reviewer: "to store a file system that can be accessed from the DUT".

These are utilized to both interact with the DUT (with the UART) and load in Linux (with the SDCard).
The following steps describe how to build and run buildroot Linux on the prototype platform.

Building Linux with FireMarshal
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Since the prototype currently does not have a block device setup for it, we build Linux with the rootfs built into the binary (otherwise known as "initramfs" or "nodisk" version of Linux).
To make building this type of Linux binary easy, we will use the FireMarshal platform (see :ref:`fire-marshal` for more information).

1. Setup FireMarshal (see :ref:`fire-marshal` on the initial setup).
2. By default, FireMarshal is setup to work with FireSim.
   Instead, we want to target the prototype platform.
   This is done by switching the FireMarshal "board" from "firechip" to "prototype" using ``marshal-config.yaml``:

.. code-block:: shell

    # this assumes you do not have a `marshal-config.yaml` file already setup
    echo "board-dir : 'boards/prototype'" > $PATH_TO_FIREMARSHAL/marshal-config.yaml

.. Note:: Refer to the FireMarshal docs on more ways to set the board differently through environment variables and more.

> 3. Next, build the workload (a.k.a buildroot Linux) with nodisk with FireMarshal.

Reviewer 2: ```suggestion
3. Next, build the workload (a.k.a buildroot Linux) in FireMarshal with the ``nodisk`` option flag.
```

   For the rest of these steps, we will assume you are using the base ``br-base.json`` workload.
   This workload has basic support for GPIO and SPI drivers (in addition to the default UART driver) but you can build off it in different workloads (refer to FireMarshal docs on workload inheritance).

.. code-block:: shell

    ./marshal -v -d build br-base.json # here the -d indicates --nodisk or initramfs

> .. Note:: Using the "board" FireMarshal functionality allows any child workload depending on ``br-base.json`` to use the "prototype" ``br-base.json`` rather than the FireChip version.

Reviewer 2: ```suggestion
.. Note:: Using the "board" FireMarshal functionality allows any child workload depending on the ``br-base.json`` workload specification to target a "prototype" platform rather than FireChip platform.
```

   Thus, you can re-use existing workloads that depend on ``br-base.json`` on the prototype platform by just changing the "board"!

4. The last step to generate the proper binary is to flatten it.

>    This is done by using FireMarshal's install feature and will produce a ``*-flat`` binary in the ``$PATH_TO_FIREMARSHAL/images`` directory (in our case ``br-base-bin-nodisk-flat``).

Reviewer 2: is this true? wouldn't the image and binary be created in ``$PATH_TO_FIREMARSHAL/images`` right after the ``build`` command? why do you need to install?

Reviewer 2: oh, does the install command at the ``-flat``?

Author: Correct, the binary/image are created at 1st (technically just the binary with `nodisk`), then the `install` flattens the binary so it can be `dd`'ed onto the SDCard

Author: Added a bit of clarification.

.. code-block:: shell

   ./marshal -v -d install -t prototype br-base.json

Setting up the SDCard
~~~~~~~~~~~~~~~~~~~~~

These instructions assume that you have a spare uSDCard that can be loaded with Linux and other files using two partitions.
The 1st partition will be used to store the Linux binary (created with FireMarshal or other means) while the 2nd partition will store a file system that can be accessed from the DUT.

> Additionally, these instructions assume you are using Linux with ``sudo`` privileges and ``gdisk`` but you can follow a similar set of steps on Mac (using ``gpt`` or another similar program).

Reviewer 2: ```suggestion
Additionally, these instructions assume you are using Linux with ``sudo`` privileges and ``gdisk``, but you can follow a similar set of steps on Mac (using ``gpt`` or another similar program).
```

1. Wipe the GPT on the card using ``gdisk``.
   Use the `z` command to zap everything.
   For rest of these instructions, we assume the SDCard path is ``/dev/sdc`` (replace this with the path to your SDCard).

.. code-block:: shell

    sudo gdisk /dev/sdc

> 2. The VCU118 bootrom assumes that the Linux binary to load into memory will be located on sector 34 of the SDCard.

Reviewer 2: would it be possible to give prompt examples? (..code-block:: shell ....)

Author: I think that is a bit excessive and unnecessary. These instructions although are setup for `gdisk` are also a bit abstract to allow others to port to other disk programs. If people get stuck on this step we can add more details if need be (I presume this will be the easiest step with the FireMarshal setup being the trickiest).

   Change the default partition alignment to `1` so you can write to sector `34`.
   Do this with the `l` command.

3. Create the new GPT with `o`.
   Click yes on all the prompts.

4. Create a 512MiB partition to store the Linux binary (this can be smaller but it must be larger than the size of the Linux binary).
   Use `n` and select sector 34, with size `+1048576` (corresponding to 512MiB).

>    For the type search for the `apfs` type and use the hex number given.

Reviewer 2: ```suggestion
   For the type, search for the `apfs` type and use the hex number given.
```

5. Create a second partition to store any other files with the rest of the SDCard.
   Use `n` and use the defaults for starting sector and overall size (expand the 2nd partition to the rest of the SDCard space).

>    For the type search for the `hfs` and use the hex number given.

Reviewer 2: ```suggestion
   For the type, search for the `hfs` and use the hex number given.
```

6. Write the changes using `w`.

7. Setup the filesystem on the 2nd partition.
   Note that the ``/dev/sdc2`` points to the 2nd partition.
   Use the following command:

.. code-block:: shell

    sudo mkfs.hfs -v "PrototypeData" /dev/sdc2

Transfer and Run Linux from the SDCard
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

After you have a Linux boot binary and the SDCard is setup properly (1st partition at sector 34), you can transfer the binary to the 1st SDCard partition.
In this example, we generated a ``br-base-bin-nodisk-flat`` from FireMarshal and we will load it using ``dd``.
Note that ``sdc1`` points to the 1st partition (remember to change the ``sdc`` to your own SDCard path).

.. code-block:: shell

    sudo dd if=$PATH_TO_FIREMARSHAL/br-base-bin-nodisk-flat of=/dev/sdc1

If you want to add files to the 2nd partition, you can also do this now.

After loading the SDCard with Linux and potentially other files, you can program the FPGA and plug in the SDCard.
To interact with Linux via the UART console, you can connect to the serial port (in this case called ``ttyUSB1``) using something like ``screen``:

.. code-block:: shell

   screen -S FPGA_UART_CONSOLE /dev/ttyUSB1 115200

> Once connected you should see the binary being loaded as well as Linux output (in some cases you might need to reset the DUT).

Reviewer 2: ```suggestion
Once connected, you should see the binary being loaded as well as Linux output (in some cases you might need to reset the DUT).
```
